packages/tpetra/core/example/Additive-Schwarz-Halo/AdditiveSchwarzHalo.cpp

```diff
@@ -68,128 +61,113 @@ int main (int argc, char *argv[])
   using Teuchos::Array;
   using Teuchos::tuple;
 
-  // MPI / Kokkos initialization 
-  Teuchos::GlobalMPISession mpiSession(&argc, &argv, NULL);
-  Kokkos::initialize(argc, argv);
-  RCP<const Teuchos::Comm<int> > comm = Teuchos::DefaultComm<int>::getComm();
+  // Tpetra::ScopeGuard initializes MPI and Kokkos, if they haven't
+  // been initialized already.  Its destructor will finalize MPI and
+  // Kokkos.
+  Tpetra::ScopeGuard tpetraScope (&argc, &argv);
+  {
```

Comment:
I think you can get away without using an inner scope: destructors are called in reverse order of constructors, so the ScopeGuard should be destroyed after everything else even if they are all in the top level scope.

Reply:
@ibaned Good point, though I've heard of some compilers not implementing destruction order quite right.  Even if this is not strictly necessary, I want users to get the sense that "Tpetra stuff can only happen in this scope."